rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,823 @@
+testing, etc.
+
+The `Configurable` derive macro would form the basis for compliance. While it would generally
+provide the scaffolding to generate an implementation of the `Configurable` trait -- walking each
+field, gathering attributes and doc comments and so on -- it would also be able to validate that
+those things exists at compile time. As an example, we can enforce that all `Configuration`
+implementors are fully documented: the type itself, their subfields, and so on. It would then become
+extremely hard for developers to add new fields or types in configuration that weren't documented
+upon cutting a new Vector release.
+
+#### 2. `Configurable` trait as a vehicle for encapsulating all facets of configuration
+
+As discussed in point 1, the `Configurable` trait is meant to provide a common interface for
+configuration types, and the types used within those types, such as types from the standard library
+or third-party crates, such that they can describe their "shape", value constraints, and any other
+relevant metadata. Below is an abbreviated version of the `Configurable` trait, along with
+supporting types:
+
+```rust
+/// The shape of the field.
+///
+/// This maps similiar to the concept of JSON's data types, where types are generalized and have
+/// generalized representations. This allows us to provide general-but-relevant mappings to core
+/// types, such as integers and strings and so on, while providing escape hatches for customized
+/// types that may be encoded and decoded via "normal" types but otherwise have specific rules or
+/// requirements.
+///
+/// Additionally, the shape of a field can encode some basic properties about the field to which it
+/// is attached. For example, numbers can be bounded on or the lower or upper end, while strings
+/// could define a minimum length, or even an allowed pattern via regular expressions.
+///
+/// In this way, they describe a more complete shape of the field than simply the data type alone.
+#[derive(Clone)]
+pub enum Shape {
+    Null,
+    Boolean,
+    String(StringShape),
+    Number(NumberShape),
+    Array(ArrayShape),
+    Map(MapShape),
+    Composite(Vec<Shape>),
+}
+
+#[derive(Clone, Default)]
+pub struct StringShape {
+    minimum_length: Option<usize>,
+    maximum_length: Option<usize>,
+    allowed_pattern: Option<&'static str>,
+}
+
+#[derive(Clone)]
+pub enum NumberShape {
+    Unsigned {
+        effective_lower_bound: u128,
+        effective_upper_bound: u128,
+    },
+    Signed {
+        effective_lower_bound: i128,
+        effective_upper_bound: i128,
+    },
+    FloatingPoint {
+        effective_lower_bound: f64,
+        effective_upper_bound: f64,
+    }
+}
+
+#[derive(Clone)]
+pub struct ArrayShape {
+    element_shape: Box<Shape>,
+    minimum_length: Option<usize>,
+    maximum_length: Option<usize>,
+}
+
+#[derive(Clone)]
+pub enum MapShape {
+    Fixed(HashMap<&'static str, Shape>),
+    Dynamic(Box<Shape>),
+}
```

Comment:
We have some configuration objects that have both a fixed and dynamic portion. This too might work better as a struct with optional parts.

Reply:
What configurations are those?

Reply:
I see it in `gcp_stackdriver_logs` sink:
```rust
pub struct StackdriverResource {
    #[serde(rename = "type")]
    pub type_: String,
    #[serde(flatten)]
    pub labels: HashMap<String, Template>,
}
```

Reply:
Ah, flattened map.

Yeah, this needs to change to "required" fields and then whether or not unknown fields are allowed. Good catch.

Reply:
I updated `MapShape` in c9b543cd0 such that it now treats required fields as their own thing, and has the ability to say whether or not other yet-unnamed fields can exist, and what shape they must satisfy.